Implement polymorphism for CAN driver

AGlass0fMilk opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

This PR introduces experimental polymorphism for the `CAN` C++ API. This allows alternate implementations (eg: external CAN interfaces) that are compatible with the internal implementation.

See #13209 for the surrounding discussion.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None, it only affects the user if they have the `EXPERIMENTAL_API` feature enabled.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->
None

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@kjbracey-arm @pan- @donatieng 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

AGlass0fMilk committed:
Implement polymorphism for CAN

pan- commented on drivers/include/drivers/interfaces/InterfaceCAN.h:
```diff
@@ -0,0 +1,171 @@
+
+namespace interface {
+
+namespace can {
```
That won't play well with code using the enum values: `CAN::Reset` won't be found as a member for example. As a workaround, this can be wrapped inside a structure instead of a namespace and `interface::Can` can inherit from it .  If polymorphism is disabled then `mbed::Can` inherits from it too. 

```c++
struct can { 

enum Mode {
// enum definition
};

// Prevent slicing and user creation of base class.
protected:
    can() = default;
    ~can() = default;
    can(const can&) = default;
    can& operator=(const can&) = default;
};
```

pan- reviewed (dismissed):
Looks good, I would like to preserve behavior regards to enum in the original class. 
Please review my solution.

AGlass0fMilk committed:
Enable inheritance of CAN enum types

This commit changes the `interface::can` namespace to a `struct`. This allows the enum types to be inherited and prevents breaking old code relying on referencing eg: `CAN::RxIrq`.

When enabled, the polymorphic CAN interface class inherits from this `interface::can` struct. If not enabled, the `mbed::CAN` class inherits from `interface::can` directly.

Co-authored-by: Vincent Coubard <vincent.coubard@arm.com>

AGlass0fMilk committed:
Make interface::CAN buildable on targets without DEVICE_CAN

This commit adds provisions to enable using interface::CAN on targets that don't have DEVICE_CAN set to 1 (ie: they don't normally have a CAN peripheral).

AGlass0fMilk committed:
Fix astyle CI failures

pan- commented on drivers/include/drivers/CAN.h:
```diff
@@ -115,7 +37,13 @@ class CANMessage : public CAN_Message {
-class CAN : private NonCopyable<CAN> {
+class CAN
+#ifdef FEATURE_EXPERIMENTAL_API
+    final : public interface::CAN, private NonCopyable<CAN>
```
nit: Maybe we can drop the `NonCopyable` inheritance and replace it with deleted copy constructor and copy assignment operator. 
This can be put at the `interface::can` level instead of the default for these two operations.

AGlass0fMilk replied:
So the deleted copy constructor and copy assignment operators are inherited? I wasn't completely sure.

pan- replied:
Default generation of these special functions will depends on the availability in the base type and fields. 
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rc-zero

AGlass0fMilk replied:
I have implemented this in d39a416

AGlass0fMilk committed:
Replace NonCopyable in CAN inheritance.

This commit moves the deletion of copy constructor and copy assignment operators to the `mbed::interface::can` class, where both `mbed::CAN` and `mbed::interface::CAN` inherit enum types from. This allows `NonCopyable` to be removed from the inheritance list.
